docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+    - If we don't allow both old collations and new collations exist in a TiDB cluster(the Option 4, see below), trying to make replications/backup & recovery between old/new clusters will break this prerequisite.
+
+#### Option 1
+
+Add a series of new collatins named with the suffix "`_np_bin`"(meaning "NO PADDING BINARY"), for example, `utf8mb4_np_bin`. Such new collations don't exist in MySQL, but they're the "real ones" used by current TiDB. After upgrading to newer TiDB versions, all old collations are shown as "`_np_bin`", MySQL collations behave the same with MySQL.
+
+Pros: Requirement 1.a, 2, 3 and 4 are met.
+
+Cons: Requirement 1.b, 1.c are not met.
+
+#### Option 2
+
+Keep all collations defined in TiDB as what they were, define a series of new collations that are actually compatible with MySQL collations. For example, we can create a new `tidb_utf8_mb4_general_ci` that is the same as `utf8mb4_general_ci` in MySQL. When TiDB users want the "real" collations, they can modify their `CREATE TABLE` statements to use the new ones.
+
+Pros: Requirement 1.a, 1.b, 2 and 3 are met.
+
+Cons: Requirement 1.c and 4 are not met, the namings are also confusing: collations that are named from MySQL are different from MySQL, but the ones named different from MySQL behave the same as MySQL.
+
+#### Option 3
+
+Add no new collation, tag old collations with syntax comment. For a upgraded TiDB cluster, add comment like "`/* AS_BINARY_COLLATION */`" for the old collations; all collations with "`/* AS_BINARY_COLLATION */`" comment are treated as the old ones.
+
+
+Pros: Requirement 1.a, 2, 3 and 4 are met.
+
+Cons: Requirement 1.b and 1.c are not met.
+
+#### Option 4
+
+Only TiDB clusters that initially boostraped with the new TiDB version are allowed to enable the new collations. For old TiDB clusters, everything remains unchanged after the upgrade.
+
+We can also provide a configuration entry for the users to choose between old/new collations when deploying a new clusters.
+
+Pros: Requirement 1 and 2 are met.
+
+Cons: Since existing TiDB cluster can't get new collations enabled, requirement 3 is eliminated; Requirement 4 is not met.
+
+## Implementation
+
+All strings in MySQL are with implicit collations, if they're not explicitly specified, so we may need to check/update all codes related to strings comparison/lookup:
+
+  - Encoding/Decoding: update codec related functions in `codec`/`tablecodec` package.
+  - SQL Runtime: update like/regex/string comparison related functions in `expression` package.
+  - SQL Optimizer: If option 1 for encodings is chosen, the optimizer should build the extra  table lookup plans.
+  - DDL/Schema: check if the collation is old/new from the versions of tables/columns.
+  - Misc
+      + update string comparison related functions in `util` package.
+      + check if range functions defined in table partitions work with collations.
```

Comment:
It would be nice if this explicitly described which collation features will be supported by the framework. For example:

- Primary Weight i.e. character (yes)
- Secondary Weight i.e. accent (current proposal says a shortcut will be implemented, as a boolean for sensitivity)
- Tertiary Weight, i.e. case (current proposal says a shortcut will be implemented, as a boolean for sensitivity)
- Quaternary Weight, i.e. punctuation weight (not supported)
- PAD / NOPAD (maybe)
- Reorder and inheritance rules (??)

The MySQL blog posts give a good overview of what these features are:
http://mysqlserverteam.com/new-collations-in-mysql-8-0-0/
http://mysqlserverteam.com/mysql-8-0-1-accent-and-case-sensitive-collations-for-utf8mb4/
http://mysqlserverteam.com/mysql-8-0-collations-the-devil-is-in-the-details/

Reply:
Thank you, I will try to add those features in this proposal.

Since many of TiDB's users expect to migrate from MySQL without any upper-level change, the basic principle is that TiDB provides as much compatibility as possible with MySQL, which is, we have to pay the price for the historical burden of MySQL(like `utf8` and `utf8mb4`).

Reply:
> PAD / NOPAD

This will be supported, we've some issues related to it already, like #13664 

> Reorder and inheritance rules (??)

Currently, we don't have a plan for this.